ceno_host/src/lib.rs

```diff
@@ -0,0 +1,250 @@
+// See `make_stdin` and `consume` for the main entry points, and how this would look
+// for the host and guest respectively for the user of our library.
+// Everything else in here would be hidden.
+
+use itertools::izip;
+use rkyv::{
+    Portable, Serialize,
+    api::high::{HighSerializer, HighValidator},
+    bytecheck::CheckBytes,
+    rancor::Error,
+    ser::allocator::ArenaHandle,
+    to_bytes,
+    util::AlignedVec,
+};
+
+#[derive(Default)]
+pub struct CenoStdin {
+    pub items: Vec<AlignedVec>,
+}
+
+pub struct SerialisedCenoStdin(pub AlignedVec);
+
+impl CenoStdin {
+    pub fn write_slice(&mut self, bytes: AlignedVec) {
+        self.items.push(bytes);
+    }
+
+    pub fn write(
+        &mut self,
+        item: &impl for<'a> Serialize<HighSerializer<AlignedVec, ArenaHandle<'a>, Error>>,
+    ) -> Result<(), Error> {
+        let bytes = to_bytes::<Error>(item)?;
+        self.write_slice(bytes);
+        Ok(())
+    }
+
+    pub fn finalise(&self) -> SerialisedCenoStdin {
+        // TODO: perhaps don't hardcode 16 here.
+        // It's from rkyv's format, so we can probably take it from there somehow?
+        // TODO: clean this up.
+        let initial_offset = (size_of::<u32>() * self.items.len()).next_multiple_of(16);
+        println!("offset: {}", initial_offset);
+        let offsets: Vec<u32> = self
+            .items
+            .iter()
+            .scan(initial_offset, |acc, bytes| {
+                let output = (*acc + bytes.len()) as u32;
+                print!("len: {}\t", bytes.len());
+                *acc += bytes.len().next_multiple_of(16);
+                println!("acc: {}", *acc);
+                Some(output)
+            })
+            .collect();
+        let offsets_u8: Vec<u8> = offsets.iter().copied().flat_map(u32::to_le_bytes).collect();
+        let mut buf: AlignedVec = AlignedVec::new();
+        buf.extend_from_slice(&offsets_u8);
+        println!("buf.len() after offsets: {}", buf.len());
+        buf.extend_from_slice(&vec![0; buf.len().next_multiple_of(16) - buf.len()]);
+        println!("buf.len() after offset padding: {}", buf.len());
+        for (offset, item) in izip!(offsets, &self.items) {
+            buf.extend_from_slice(item);
+            buf.extend_from_slice(&vec![0; buf.len().next_multiple_of(16) - buf.len()]);
+            assert_eq!(buf.len(), offset.next_multiple_of(16) as usize);
+        }
+        SerialisedCenoStdin(buf)
+    }
+}
+
+pub struct SerialisedCenoStdinIter<'a> {
+    buf: &'a SerialisedCenoStdin,
+    next: usize,
+}
+
+impl<'b> SerialisedCenoStdinIter<'b> {
+    pub fn read<'a, T>(&'a mut self) -> &'b T
+    where
+        T: Portable + for<'c> CheckBytes<HighValidator<'c, Error>>,
+    {
+        rkyv::access::<T, Error>(self.read_slice()).unwrap()
```

Comment:
if I remember correctly, from your previous example, we can throw the whole deliberate slice into rkyv and iterator through it, without implementing a wrapping iterator by ourself?

Reply:
Yes, in principle.  But that only works, if we make sure that the end of the mapped data always shows up at a consistent fixed memory address.  `rkyv` finds its data from the end.

Since we decided that our VM can only put the start of our mapped data at a fixed consistent address, we would need to eg store the length of the mapped data at the fixed start of the region.

But once we are mucking around with bytes ourselves, we might as well do the whole outer layer by ourselves, since it's very simple.

There's some additional complexity required for directly storing rkyv-serialised data inside of rkyv in such a way that you can zero-copy deserialise both layers.  The extra complexity is mostly around the 16 byte alignment requirement.  That's why my original prototype used the `unaligned` feature.

It's totally doable, but then probably not really simpler than coding up this outer layer ourselves.

However, I'm open to using rkyv for the outer layer as well, if you think that's cleaner.  The code in here is just a prototype.